### DISCUSSION TITLE
IPOPT Mex File Compilation with MATLAB MA57 Library

### DISCUSSION THREAD

**Discussion by anilvrao** [coin-or/Ipopt#764] [Q&A]
Dear All,

I continue to grapple with the ability to successfully compile an IPOPT Mex file for Mac (Intel or Apple Silicon) using the built-in MATLAB MA57 library.  I have tried so many things and nothing seems to work.  Let me see if I can explain from the top.  

First, I used a script written by Josh Harris (thank you Josh) to compile IPOPT by downloading all of the necessary components and then completing the compilation.  The script is attached to this message.  Here is the configure line in that script (which I modified to include "--enable-matlab-ma57" and "--without-hsl":

    "$IPOPTDIR/configure" --prefix="$COMPILE_IPOPT_INSTALL_PREFIX" --enable-matlab-ma57 --enable-static=yes --enable-shared=yes --disable-java --disable-pardisomkl --without-asl --without-hsl

Thus, while I believe that the MATLAB MA57 library is being enabled when compiling IPOPT, I am not sure if what I suggest is actually true.  

Next, after compiling IPOPT, I have added the following compiler flags in the script CompileIpoptMexLib.m supplied by Enrico Bertolazzi (thank you, Enrico):  

-L/usr/local/lib -L/Applications/MATLAB_R2024a.app/bin/maca64 -lmwma57

Thus, again, I am asking the compiler to include the MATLAB directory that contains the library libmwma57.dylib.  Interestingly, I do to get an error when running CompileIpoptMexLib.m, which leads me to think that the MATLAB MA57 library is being found during the running of CompileIpoptMexlib.m.  

Finally, after all is said and done the Mex file is compiled.  When I attempt to run the Mex file, however, I receive the following error:

Exception of type: DYNAMIC_LIBRARY_FAILURE in file "/Users/anilvrao/Documents/MATLAB/harrisIpopt/Ipopt/src/Common/IpLibraryLoader.cpp" at line 86:
 Exception message: dlopen(libhsl.dylib, 0x0002): tried: 'libhsl.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OSlibhsl.dylib' (no such file), '/opt/homebrew/opt/libomp/lib/libhsl.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/opt/libomp/lib/libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/./libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/../../sys/os/maca64/libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/../../runtime/maca64/libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/../../sys/java/jre/maca64/jre/lib/server/libhsl.dylib' (no such file), '/Applications/MATLAB_R2024a.app/bin/maca64/../../sys/java/jre/maca64/jre/lib/libhsl.dylib' (no such file), '/usr/lib/libhsl.dylib' (no such file, not in dyld cache), 'libhsl.dylib' (no such file)

Of course, I am wondering why the Mex file is looking for the library "libhsl.dylib" when I have specifically asked to enable the MATLAB MA57 library and have included all of the necessary paths during compilation.  

I am not sure what is going on.  Please see the shell script attached, and please rename it compile_ipopt.sh before trying to use it to perform the compilation.  

Thanks!

-- Anil

[compile_ipopt.txt](https://github.com/coin-or/Ipopt/files/15227468/compile_ipopt.txt)

**Comment by svigerske** [comment]
We went through this in #722 already. There is no `--enable-matlab-ma57` option.

Just because you link Ipopt to mwma57, it won't know that MA57 is available. When compiling the Ipopt source, it needs to be known that one will link against an HSL library. For this purpose, there are the `--with-hsl-cflags` and `--with-hsl-lflags` options. Otherwise, Ipopt will try to load an HSL library at runtime. Often, this is also just fine, and you would have to set option [hsllib](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_hsllib) to the path of Matlab's HSL library.

However, there is always the problem that Ipopt assumes 32bit integers, while Matlab, including its HSL library, uses 64bit integers. For that reason, you can define `FUNNY_MA57_INT` to have Ipopt's MA57 interface use 64bit integers. Still, there could be a problem with Lapack libraries, since both Ipopt and Lapack use them, and one assumes 32bit and the other 64bit integers. Linking statically against a Lapack library seems to help.
